src/units/drawer.cpp

```diff
@@ -462,6 +469,39 @@ void unit_drawer::redraw_unit(const unit& u) const
 		halo_man.set_location(ac.unit_halo_, halo_x, halo_y);
 	}
 
+	const std::vector<std::string> halos_abilities = u.halo_abilities();
+	bool has_abil_halo = !ac.abil_halos_.empty() && ac.abil_halos_.front() && ac.abil_halos_.front()->valid();
```

Comment:
Does checking `ac.abil_halos_.front()` do anything? If it's not empty then you already know there's at least a single element in the vector.

Reply:
It's because ->valid what i use .front and not all vector

Reply:
corrected in https://github.com/wesnoth/wesnoth/pull/7856/commits/e64453540085014bc53651ad64b311fd61d641de, now abil_halos_vector contain only valid halos.

Reply:
What I meant was why it was it `!ac.abil_halos_.empty() && ac.abil_halos_.front() && ac.abil_halos_.front()->valid()` instead of just `!ac.abil_halos_.empty() && ac.abil_halos_.front()->valid()`. Is it possible for elements in the vector to be `nullptr`?

Reply:
i don't know, because i don't know how abil_halos_.front() could be invalid, but if it is case but other lement are valid then using the code in https://github.com/wesnoth/wesnoth/commit/e64453540085014bc53651ad64b311fd61d641de will be more adequat if vector can contin what valid halo, verify what it is empty() or not is sufficient.